File: freq.py
import sys
import re

assert(len(sys.argv) == 4, "Wrong number of arguments. Syntax is freq [book] [stopwords] [n]")
try: 
    n = int(sys.argv[3])
except:
    raise TypeError("n must be an integer")

stopwords_path = sys.argv[2]
book_path = sys.argv[1]

def make_stopword_dict(filename):
    stopwords = set()
    with open(filename, "r") as f:
        s = f.read()
        stopwords.update(s.split(","))
    
    return stopwords

def calc_occurences(filename, stopwords):
    with open(filename, "r") as f:
        s = f.read()
        wordslist = re.split("[^\w]", s.lower())
        finaldict = {}
        for w in wordslist: 
            if len(w) > 1 and w not in stopwords:
                finaldict.setdefault(w, 0)
                finaldict[w] += 1
        return finaldict

stopwords = make_stopword_dict(stopwords_path)

occurences = calc_occurences(book_path, stopwords)

for (k, (i, j)) in enumerate(sorted(occurences.items(), key = lambda x: x[1], reverse=True)[:n]):
    print(f"""{k+1}: "{i}", with {j} occurences""")

File: freq.ml
module StringMap = Map.Make(String)
module StringSet = Set.Make(String)

let is_letter (c: Char.t) = (Char.code c) >= (Char.code 'a') && (Char.code c) <= (Char.code 'z')

let make_stopwords ic = 
	let str = In_channel.input_all ic in
    let str = String.lowercase_ascii str in
	let wordslist = String.split_on_char ',' str in
	let occurences : StringSet.t = StringSet.empty in
    List.fold_right StringSet.add wordslist occurences

let make_occurences (stopwords: StringSet.t) ic = 
    let str = In_channel.input_all ic in
    let str = String.lowercase_ascii str in
    let remove_seps = function 
		|c when is_letter c -> c
      	|_ -> ' ' in
    let str = String.map remove_seps str in
    let wordslist = String.split_on_char ' ' str in
	let wordslist = List.filter (fun v -> String.length v >= 2 && not (StringSet.mem v stopwords)) wordslist in
    let occurences : int StringMap.t = StringMap.empty in
    let add_to_map map v = StringMap.update v (function None -> Some 1 | Some x -> Some (x+1)) map in
    let occurences = List.fold_left add_to_map occurences wordslist in
    List.fast_sort (fun (_, i1) (_, i2) -> Int.compare i2 i1) (StringMap.to_list occurences)
  
let () = if Array.length Sys.argv != 4 then failwith "Wrong number of arguments. Syntax is freq [book] [stopwords] [n]"
    else let n = int_of_string_opt Sys.argv.(3) in 
	match n with 
		|None -> failwith "n must be an integer"
		|Some n ->
			let stopwords = In_channel.with_open_text Sys.argv.(2) make_stopwords in
			let occurences = In_channel.with_open_text Sys.argv.(1) (make_occurences stopwords) in 
			Seq.iteri (fun i (a, b) -> Printf.printf "%d: \"%s\", with %d occurences\n" i a b) (Seq.take n (List.to_seq occurences))
